### PR TITLE
Add events around type load and r2r usage

### PR DESCRIPTION
**Pull request by davidwrighton**
Add events to track the time spent loading types, and making R2R code useable
- Allows for investigation of type loader performance
  - Uses a new keyword, as this event will generate many strings which is itself fairly slow
- Changes R2R events from Verbose events to Informational events
  - These events are already controlled by a separate keyword, and don't need to also be only emitted when Verbose eventing is on

### CODE REVIEW

**Comment by davidwrighton** [comment]
@mangod9 @billwert

**Comment by mjsabby** [comment]
@davmason how are schemas changes for this delivered to event pipe consumers? Does the Event Pipe infrastructure read the .man files?

**Review by jkotas — APPROVED** [review]
Thanks!

**Comment by davmason** [comment]
@mjsabby EventPipe will create an event for the native runtime events that has the name and provider but the metadata will always say it has no parameters. The payload will be there for each event that is fired though, so you have to hard code event parsing for the native runtime events.

**Comment by mjsabby** [comment]
Thanks. I wonder if carrying the manifest file could be useful in the runtime. I guess it's not that a big deal on shipped runtimes cause you can always look up the shipped one and use that. During these preview periods, you have to make an effort to keep it in sync if you have an out of band mechanism which I guess I will need.

File: src/coreclr/src/vm/ClrEtwAll.man [hunk 1]
```diff
@@ -3364,11 +3421,16 @@
                            task="CLRMethod"
                            symbol="MethodLoad_V2" message="$(string.RuntimePublisher.MethodLoad_V2EventMessage)"/>
 
-                    <event value="159" version="0" level="win:Verbose" template="R2RGetEntryPoint"
+                    <event value="159" version="0" level="win:Informational" template="R2RGetEntryPoint"
                            keywords ="CompilationDiagnosticKeyword" opcode="MethodLoad"
                            task="CLRMethod"
                            symbol="R2RGetEntryPoint" message="$(string.RuntimePublisher.R2RGetEntryPointEventMessage)"/>
 
+                    <event value="160" version="0" level="win:Informational" template="R2RGetEntryPointStarted"
```
Comment (noahfalk) [inline src/coreclr/src/vm/ClrEtwAll.man:3429]:
```suggestion
                    <event value="160" version="0" level="win:Informational" template="R2RGetEntryPointStart"
```

Convention for ETW event naming is to use "Start" and "Stop" (in EventSource the naming triggers a specific behavior that pairs start/stop events, in our native events it has no behavioral effects but its nice to match the convention)

File: src/coreclr/src/vm/eventtrace.cpp [hunk 2]
```diff
@@ -3152,6 +3153,69 @@ CrstBase * ETW::TypeSystemLog::GetHashCrst()
...
+UINT32 s_TypeLoadOps = 0;
+
+UINT32 ETW::TypeSystemLog::TypeLoadBegin()
+{
+    CONTRACTL{
+        NOTHROW;
+        GC_TRIGGERS;
+    } CONTRACTL_END;
+
+    UINT32 typeLoad = InterlockedIncrement(&s_TypeLoadOps);
```
Comment (noahfalk) [inline src/coreclr/src/vm/eventtrace.cpp:3168]: Is there any value to incrementing this counter when we aren't tracking the events? May as well move the Increment inside the if body.

File: src/coreclr/src/vm/eventtrace.cpp [hunk 3]
```diff
@@ -3152,6 +3153,69 @@ CrstBase * ETW::TypeSystemLog::GetHashCrst()
...
+
+    if (ETW_EVENT_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context, TypeLoadStart))
+    {
+        EX_TRY
```
Comment (noahfalk) [inline src/coreclr/src/vm/eventtrace.cpp:3172]: EH shouldn't be needed here

**Review by noahfalk — APPROVED** [review]
LGTM, some nits on naming and perf inline

**Comment by noahfalk** [comment]
> EventPipe will create an event for the native runtime events that has the name and provider but the metadata will always say it has no parameters

Interesting, I hadn't realized that had never been implemented. As @mjsabby suggested we should probably keep a lookout if the out-of-band updates to TraceEvent are causing too much labor/confusion. It probably wouldn't be that hard to pregenerate the metadata given that we already had to understand each event signature to generate the NativeRuntimeEventSource source code.

File: src/coreclr/src/vm/eventtrace.cpp (hunk 2 above)
Comment (davidwrighton) [reply src/coreclr/src/vm/eventtrace.cpp]: This is something that I've wanted for a while as a performance counter like thing. I don't have good evidence as to whether or not its really all that useful, but this should be enough to see the value. Its sufficiently cheap that I would just like to enable it unconditionally.

File: src/coreclr/src/vm/eventtrace.cpp [hunk 4]
```diff
@@ -5305,6 +5369,26 @@ VOID ETW::MethodLog::GetR2RGetEntryPoint(MethodDesc *pMethodDesc, PCODE pEntryPo
...
+VOID ETW::MethodLog::GetR2RGetEntryPointStarted(MethodDesc *pMethodDesc)
+{
+    CONTRACTL{
+        NOTHROW;
+        GC_TRIGGERS;
+    } CONTRACTL_END;
+
+    if (ETW_EVENT_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context, R2RGetEntryPointStarted))
+    {
+        EX_TRY
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/coreclr/src/vm/eventtrace.cpp:5381]: Ah, copy paste issue from the R2RGetEntrypointEvent stuff, where it is necessary as there is more than just a fire entrypoint.

File: src/coreclr/src/vm/readytoruninfo.cpp [hunk 5]
```diff
@@ -830,6 +830,11 @@ PCODE ReadyToRunInfo::GetEntryPoint(MethodDesc * pMD, PrepareCodeConfig* pConfig
     if (m_readyToRunCodeDisabled)
         goto done;
 
+    if (ETW_EVENT_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context, R2RGetEntryPointStarted))
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/coreclr/src/vm/readytoruninfo.cpp:833]: No, I'll remove the existing one for the R2RGetEntryPoint event too.

File: src/coreclr/src/vm/eventtrace.cpp [hunk 6]
```diff
@@ -5305,6 +5369,26 @@ VOID ETW::MethodLog::GetR2RGetEntryPoint(MethodDesc *pMethodDesc, PCODE pEntryPo
     }
 }
 
+VOID ETW::MethodLog::GetR2RGetEntryPointStarted(MethodDesc *pMethodDesc)
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/coreclr/src/vm/eventtrace.cpp:5372]: The Get isn't superfluous. There is a naming conflict if the name if the MethodLog method actually matches the event name. For the other event this is paired with, a "Get" prefix was added to make the thing compile. I've followed the same naming adjustment. OTOH, Started vs Start... I've changed to Start.

File: src/coreclr/src/vm/readytoruninfo.cpp [hunk 7]
```diff
@@ -932,10 +934,7 @@ PCODE ReadyToRunInfo::GetEntryPoint(MethodDesc * pMD, PrepareCodeConfig* pConfig
     }
 
 done:
-    if (ETW_EVENT_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context, R2RGetEntryPoint))
-    {
-        ETW::MethodLog::GetR2RGetEntryPoint(pMD, pEntryPoint);
-    }
+    ETW::MethodLog::GetR2RGetEntryPoint(pMD, pEntryPoint);
```
Comment (jkotas) [inline src/coreclr/src/vm/readytoruninfo.cpp:937]: IIRC, the out-lined check here was added to fix a minor performance regression. The allocation of the huge stack-allocated buffer in the tracing helpers was showing up on the profile.

File: src/coreclr/src/vm/eventtrace.cpp (hunk 2 above)
Comment (noahfalk) [reply src/coreclr/src/vm/eventtrace.cpp]: happy to keep it as long as you find it useful : )

**Comment by mjsabby** [comment]
@noahfalk sounds good. I was going to generate the [EVENT_PROPERTY_INFO](https://docs.microsoft.com/en-us/windows/win32/api/tdh/ns-tdh-event_property_info) out of the manifest and have a lookup table for that. Do you already have some code for that?

File: src/coreclr/src/vm/ClrEtwAll.man [hunk 8]
```diff
@@ -3364,11 +3421,16 @@
                            task="CLRMethod"
                            symbol="MethodLoad_V2" message="$(string.RuntimePublisher.MethodLoad_V2EventMessage)"/>
 
-                    <event value="159" version="0" level="win:Verbose" template="R2RGetEntryPoint"
+                    <event value="159" version="0" level="win:Informational" template="R2RGetEntryPoint"
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/coreclr/src/vm/ClrEtwAll.man:3424]: Yes.

File: src/coreclr/src/vm/readytoruninfo.cpp (hunk 7 above)
Comment (davidwrighton) [reply src/coreclr/src/vm/readytoruninfo.cpp:937]: Oh that's not desirable, I'll put it back in then for that helper and for the TypeLoadStop call, no need to have unnecessary performance regressions.

**Comment by noahfalk** [comment]
> @noahfalk sounds good. I was going to generate the EVENT_PROPERTY_INFO out of the manifest and have a lookup table for that. Do you already have some code for that?

I'm not aware of anything we have that specifically converts the manifest into EVENT_PROPERTY_INFO, but we do have scripts such as [this one](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/scripts/genRuntimeEventSources.py) that reads the manifest to create other artifacts at build time.

**Review by brianrob — APPROVED** [review]

**Pull request merged**